File: python_port/rag/observability/trace_bus.py
from typing import List, Any, Protocol

class Observer(Protocol):
    def on_event(self, stage: str, inputs: Any, outputs: Any, duration_ms: int) -> None:
        ...

class TraceBus:
    _observers: List[Observer] = []

    @classmethod
    def register(cls, observer: Observer) -> None:
        cls._observers.append(observer)

    @classmethod
    def emit(cls, stage: str, inputs: Any, outputs: Any, duration_ms: int) -> None:
        for observer in cls._observers:
            observer.on_event(stage, inputs, outputs, duration_ms)

File: Iteration1CSE3063Proje/src/main/java/com/cse3063/rag/observability/TraceBus.java
package com.cse3063.rag.observability;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class TraceBus {
    public interface Observer {
        void onEvent(String stage, Object inputs, Object outputs, long durationMs);
    }

    private static final List<Observer> observers = new ArrayList<>();

    public static void register(Observer observer) {
        observers.add(observer);
    }

    public static void emit(String stage, Object inputs, Object outputs, long durationMs) {
        for (Observer observer : observers) {
            observer.onEvent(stage, inputs, outputs, durationMs);
        }
    }
}
